File: engine/razzle/core/moves.py
"""
Move generation for Razzle Dazzle.

Handles knight moves and ball passes with proper rule enforcement.
"""

from __future__ import annotations
from typing import Iterator

from .bitboard import (
    ROWS, COLS, NUM_SQUARES,
    KNIGHT_ATTACKS, RAY_MASKS, BETWEEN,
    bit, iter_bits, sq_to_rowcol, rowcol_to_sq, sq_to_algebraic,
    is_valid_sq
)
from .state import GameState


# Move encoding: src * 56 + dst
def encode_move(src: int, dst: int) -> int:
    """Encode a move as a single integer."""
    return src * NUM_SQUARES + dst


def decode_move(move: int) -> tuple[int, int]:
    """Decode a move into (src, dst)."""
    return move // NUM_SQUARES, move % NUM_SQUARES


def move_to_algebraic(move: int) -> str:
    """Convert move to algebraic notation."""
    src, dst = decode_move(move)
    return f"{sq_to_algebraic(src)}-{sq_to_algebraic(dst)}"


def algebraic_to_move(s: str) -> int:
    """Parse algebraic notation to move."""
    from .bitboard import algebraic_to_sq
    parts = s.strip().split('-')
    if len(parts) != 2:
        raise ValueError(f"Invalid move format: {s}")
    src = algebraic_to_sq(parts[0])
    dst = algebraic_to_sq(parts[1])
    return encode_move(src, dst)


class MoveGenerator:
    """Generates legal moves for a game state."""

    @staticmethod
    def get_knight_moves(state: GameState) -> Iterator[int]:
        """
        Generate all legal knight moves for current player.

        A piece can only move if it does NOT have the ball.
        """
        p = state.current_player
        pieces = state.pieces[p]
        ball_sq = None

        # Find which piece has the ball (if any)
        for sq in iter_bits(state.balls[p]):
            ball_sq = sq

        # A piece at ball position cannot move (it's holding the ball)
        movable_pieces = pieces & ~state.balls[p]

        for src in iter_bits(movable_pieces):
            # Knight attacks from this square, filtered to empty squares
            targets = KNIGHT_ATTACKS[src] & state.empty
            for dst in iter_bits(targets):
                yield encode_move(src, dst)

    @staticmethod
    def get_pass_moves(state: GameState) -> Iterator[int]:
        """
        Generate all legal ball passes for current player.

        The ball can be passed in a straight line (horizontal, vertical, diagonal)
        to any friendly piece that hasn't touched the ball this turn.
        """
        p = state.current_player
        ball_pos = state.balls[p]

        if not ball_pos:
            return

        ball_sq = next(iter_bits(ball_pos))
        my_pieces = state.pieces[p]
        occupied = state.occupied

        # Can't pass to pieces that already touched ball this turn
        valid_receivers = my_pieces & ~state.touched_mask

        # Check each direction
        directions = [
            (1, 0), (-1, 0), (0, 1), (0, -1),  # N, S, E, W
            (1, 1), (1, -1), (-1, 1), (-1, -1)  # Diagonals
        ]

        ball_row, ball_col = sq_to_rowcol(ball_sq)

        for dr, dc in directions:
            r, c = ball_row + dr, ball_col + dc
            while is_valid_sq(r, c):
                sq = rowcol_to_sq(r, c)
                sq_bit = bit(sq)

                if occupied & sq_bit:
                    # Hit something
                    if valid_receivers & sq_bit:
                        # It's our piece and can receive
                        yield encode_move(ball_sq, sq)
                    break  # Can't pass through pieces

                r += dr
                c += dc

    @staticmethod
    def must_pass(state: GameState) -> bool:
        """
        Check if current player MUST pass (opponent adjacent to ball).

        "If your opponent's previous move results in one of her pieces
        being adjacent to your ball, you must pass if you can."
        """
        p = state.current_player
        ball_pos = state.balls[p]
        if not ball_pos:
            return False

        ball_sq = next(iter_bits(ball_pos))
        ball_row, ball_col = sq_to_rowcol(ball_sq)

        opp_pieces = state.pieces[1 - p]

        # Check all 8 adjacent squares
        for dr in [-1, 0, 1]:
            for dc in [-1, 0, 1]:
                if dr == 0 and dc == 0:
                    continue
                r, c = ball_row + dr, ball_col + dc
                if is_valid_sq(r, c):
                    adj_sq = rowcol_to_sq(r, c)
                    if opp_pieces & bit(adj_sq):
                        return True

        return False

    @staticmethod
    def get_legal_moves(state: GameState) -> list[int]:
        """
        Get all legal moves for current player.

        Returns list of encoded moves. Move type is determined by context:
        - If source is ball position: it's a pass
        - Otherwise: it's a knight move

        Rules:
        1. If opponent is adjacent to your ball, you MUST pass if possible
        2. You can chain passes (each is a separate move in atomic encoding)
        3. After passing, you can move a piece OR continue passing
        4. A knight move ends your turn
        """
        moves = []

        # Check if we must pass
        forced_pass = MoveGenerator.must_pass(state)
        pass_moves = list(MoveGenerator.get_pass_moves(state))

        if forced_pass and pass_moves:
            # Must pass - only return pass moves
            return pass_moves

        # Can do either passes or knight moves
        moves.extend(pass_moves)
        moves.extend(MoveGenerator.get_knight_moves(state))

        return moves

    @staticmethod
    def get_move_mask(state: GameState) -> list[bool]:
        """
        Get a mask indicating which moves are legal.

        Returns a list of NUM_SQUARES * NUM_SQUARES booleans.
        Index i is True if move i is legal.

        Useful for neural network output masking.
        """
        mask = [False] * (NUM_SQUARES * NUM_SQUARES)
        for move in MoveGenerator.get_legal_moves(state):
            mask[move] = True
        return mask


# Convenience functions
def get_legal_moves(state: GameState) -> list[int]:
    """Get all legal moves for the current player."""
    return MoveGenerator.get_legal_moves(state)


def is_legal_move(state: GameState, move: int) -> bool:
    """Check if a move is legal."""
    return move in MoveGenerator.get_legal_moves(state)


def get_move_count(state: GameState) -> int:
    """Get number of legal moves."""
    return len(MoveGenerator.get_legal_moves(state))

File: webapp/src/engine/moves.ts
/**
 * Move generation for Razzle Dazzle.
 *
 * Handles knight moves and ball passes with proper rule enforcement.
 */

import {
  NUM_SQUARES,
  KNIGHT_ATTACKS,
  bit,
  iterBits,
  sqToRowCol,
  isValidSq,
  rowcolToSq,
  END_TURN_MOVE,
} from './bitboard';
import type { EngineState } from './state';
import { getOccupied, getEmpty } from './state';

export function encodeMove(src: number, dst: number): number {
  return src * NUM_SQUARES + dst;
}

export function decodeMove(move: number): [number, number] {
  return [Math.floor(move / NUM_SQUARES), move % NUM_SQUARES];
}

/** Generate all legal knight moves for the current player. */
export function getKnightMoves(state: EngineState): number[] {
  const moves: number[] = [];
  const p = state.currentPlayer;
  const movablePieces = state.pieces[p] & ~state.balls[p];
  const empty = getEmpty(state);

  for (const src of iterBits(movablePieces)) {
    const targets = KNIGHT_ATTACKS[src] & empty;
    for (const dst of iterBits(targets)) {
      moves.push(encodeMove(src, dst));
    }
  }
  return moves;
}

/** Generate all legal ball passes for the current player. */
export function getPassMoves(state: EngineState): number[] {
  const moves: number[] = [];
  const p = state.currentPlayer;
  const ballPos = state.balls[p];
  if (ballPos === 0n) return moves;

  // Find ball square
  let ballSq = -1;
  for (const sq of iterBits(ballPos)) {
    ballSq = sq;
    break;
  }
  if (ballSq === -1) return moves;

  const myPieces = state.pieces[p];
  const occupied = getOccupied(state);
  const validReceivers = myPieces & ~state.touchedMask;

  const directions: [number, number][] = [
    [1, 0],
    [-1, 0],
    [0, 1],
    [0, -1], // N, S, E, W
    [1, 1],
    [1, -1],
    [-1, 1],
    [-1, -1], // Diagonals
  ];

  const [ballRow, ballCol] = sqToRowCol(ballSq);

  for (const [dr, dc] of directions) {
    let r = ballRow + dr;
    let c = ballCol + dc;
    while (isValidSq(r, c)) {
      const sq = rowcolToSq(r, c);
      const sqBit = bit(sq);
      if (occupied & sqBit) {
        // Hit something - if it's our piece and can receive, it's a valid pass
        if (validReceivers & sqBit) {
          moves.push(encodeMove(ballSq, sq));
        }
        break; // Can't pass through pieces
      }
      r += dr;
      c += dc;
    }
  }

  return moves;
}

/**
 * Check if current player MUST pass (opponent moved adjacent to ball).
 */
export function mustPass(state: EngineState): boolean {
  if (state.lastKnightDst < 0) return false;

  const p = state.currentPlayer;
  const ballPos = state.balls[p];
  if (ballPos === 0n) return false;

  let ballSq = -1;
  for (const sq of iterBits(ballPos)) {
    ballSq = sq;
    break;
  }
  if (ballSq === -1) return false;

  const [ballRow, ballCol] = sqToRowCol(ballSq);
  const [dstRow, dstCol] = sqToRowCol(state.lastKnightDst);

  const rowDiff = Math.abs(ballRow - dstRow);
  const colDiff = Math.abs(ballCol - dstCol);

  return rowDiff <= 1 && colDiff <= 1 && (rowDiff > 0 || colDiff > 0);
}

/** Get all legal moves for the current player. */
export function getLegalMoves(state: EngineState): number[] {
  const forcedPass = mustPass(state);
  const passMoves = getPassMoves(state);

  if (forcedPass && passMoves.length > 0) {
    return passMoves;
  }

  if (state.hasPassed) {
    return [...passMoves, END_TURN_MOVE];
  }

  return [...passMoves, ...getKnightMoves(state)];
}

/**
 * Get a boolean mask indicating which moves are legal.
 * Returns array of length NUM_SQUARES * NUM_SQUARES (3136).
 * END_TURN (-1) is NOT included in this mask; check separately.
 */
export function getMoveMask(state: EngineState): boolean[] {
  const mask = new Array<boolean>(NUM_SQUARES * NUM_SQUARES).fill(false);
  for (const move of getLegalMoves(state)) {
    if (move >= 0) {
      mask[move] = true;
    }
  }
  return mask;
}
